plugin/main/src/kotlinx/benchmark/gradle/NativeMultiplatformTasks.kt

```diff
@@ -116,30 +118,127 @@ fun Project.createNativeBenchmarkExecTask(
+
+        val ideaActive = (extensions.extraProperties.get("idea.internal.test") as? String)?.toBoolean() ?: false
+        configFile = writeParameters(target.name, reportFile, if (ideaActive) "xml" else "text", config)
 
         dependsOn(linkTask)
         doFirst {
-            val ideaActive = (extensions.extraProperties.get("idea.internal.test") as? String)?.toBoolean() ?: false
-            args(writeParameters(target.name, reportFile, if (ideaActive) "xml" else "text", config))
             reportsDir.mkdirs()
             logger.lifecycle("Running '${config.name}' benchmarks for '${target.name}'")
         }
     }
 }
 
-open class NativeBenchmarkExec : Exec() {
+open class NativeBenchmarkExec() : DefaultTask() {
 /*
     @Option(option = "filter", description = "Configures the filter for benchmarks to run.")
     var filter: String? = null
 */
+    @Input
+    lateinit var executable: String
+
+    var workingDir: File? = null
+
+    @Input
+    lateinit var configFile: File
+
+    @Input
+    lateinit var config: BenchmarkConfiguration
+
+    @Input
+    lateinit var reportFile: File
+
+    @Input
+    lateinit var configDir: File
+
+    private fun execute(args: Collection<String>) {
+        project.exec {
+            it.executable = executable
+            it.args(args)
+            if (workingDir != null)
+                it.workingDir = workingDir
+        }
+    }
+
+    @ExperimentalPathApi
+    @TaskAction
+    fun run() {
+        // Get full list of running benchmarks
+        execute(listOf(configFile.absolutePath, "--list", configDir.absolutePath))
+        val detailedConfigFiles = project.fileTree(configDir).files.sortedBy { it.absolutePath }
+        val jsonReportParts = mutableListOf<File>()
+
+        detailedConfigFiles.forEach { runConfig ->
+            val runConfigPath = runConfig.absolutePath
+            val lines = runConfig.readLines()
+            require(lines.size > 1) { "Wrong detailed configuration format" }
+            val currentConfigDescription = lines[1]
+
+            // Execute benchmark
+            if (config.iterationMode == "internal") {
+                val jsonFile = createTempFile("bench", ".json").toFile()
+                jsonReportParts.add(jsonFile)
+                execute(listOf(configFile.absolutePath, "--internal", runConfigPath, jsonFile.absolutePath))
+            } else {
+                val iterations = currentConfigDescription.substringAfter("iterations=")
+                    .substringBefore(',').toInt()
+                val warmups = currentConfigDescription.substringAfter("warmups=")
+                    .substringBefore(',').toInt()
+                // Warm up
+                var exceptionDuringExecution = false
+                for(i in 0 until warmups) {
+                    val textResult = createTempFile("bench", ".txt").toFile()
+                    execute(listOf(configFile.absolutePath, "--warmup", runConfigPath, i.toString(), textResult.absolutePath))
```

Comment:
What is the benefit of the warmups if they are executed in their own processes?

Reply:
Warm-up for native applications are needed to warmup processors caches, etc..